fix: implicit initializers

kevinclancy opened the pull request:
### Notes

This PR adds code that implicitly inserts a slither AST for default-value initialization code during analysis.

### Testing

Use the following test file `test.sol`:
```
pragma solidity ^0.8.13;

struct A {
    int hello;
}

contract Test {
    function testFunc() external {
        A[3] memory z;
    }
}
```

* Running `pipenv run slither test.sol --print certikir` shows that default value initialization has been generated, whereas no default value initialization is generated when we run `pipenv run slither test.sol --print slithir`. 

* Run `./evaluates.sh run 100` from `tools-exec` and verify that all projects succeed.

### Related Issue
https://github.com/CertiKProject/slither-task/issues/200

Commit:
added implicit initializers

duckki commented on slither/solc_parsing/variables/variable_declaration.py:
```diff
@@ -200,6 +222,47 @@ def _init_from_declaration(self, var: Dict, init: bool):  # pylint: disable=too-
                 self._initializedNotParsed = var["children"][1]
 
+    def _get_init_value(self, ty : Type) -> Expression:
+
```
I think we need a data location check. When the data location is "storage", the variable is initialized as a null reference, instead of the default value.

```solidity
pragma solidity ^0.8.0;

contract ImplicitInitializer {

    string name;

    function storage_var() public view {
        string storage v2; // v2 is a null reference.
        v2 = name;
    }

}
```

kevinclancy replied:
There doesn't seem to be any way to represent a null reference in SlithIR. However, the solc compiler will produce an error if we try to use an uninitialized state variable. So if we refrain from generating any code at the declaration site, we can rely on it being initialized later using an assignment.

duckki commented on slither/solc_parsing/variables/variable_declaration.py:
```diff
@@ -47,6 +64,7 @@ def __init__(
         self._initializedNotParsed = None
 
         self._is_compact_ast = False
+        self._from_statement = False
```
`_from_statement` may be confusing. How about `_is_implicit_statement` (w/ inverted polarity)?

kevinclancy replied:
`_from_statement` was tracking if the Variable is local, but I decided to check that it is an instance of `LocalVariable` instead.

duckki replied:
I see. It makes sense.

duckki commented on slither/core/variables/variable.py:
```diff
@@ -22,6 +22,7 @@ def __init__(self):
         self._is_constant = False
         self._is_immutable: bool = False
         self._is_reentrant: bool = True
+        self._storage_location: Literal["memory", "storage", "calldata", "default"] = "default"
```
Actually, `LocalVariable` class already has `location` property that holds that information. Please use that.

Commit:
only generate default values for variables with 'memory' and 'default' location
check that the type of variable is LocalVariable instead of using a _from_statement variable

duckki reviewed (approved):
Looks good!

duckki commented:
BTW, I consider this is a "feat", rather than "fix", since we are not fixing an existing implicit initializer code.
